bionic/deriver.py

```diff
@@ -301,28 +304,28 @@ def register_provider(provider):
         # behavior downstream.
         self._followup_dnode_lists_by_dnode = dict(followup_dnode_lists_by_dnode)
 
-    def _set_up_bootstrap(self):
+    def _set_up_core(self):
         """
         Initializes some key objects needed to compute user-defined entities.
         """
 
-        if self._bootstrap is not None:
+        if self._core_is_finalized:
             return
 
-        self._bootstrap = Bootstrap(
-            persistent_cache=self._bootstrap_singleton_entity("core__persistent_cache"),
-            versioning_policy=self._bootstrap_singleton_entity(
-                "core__versioning_policy"
-            ),
-            aip_executor=self._bootstrap_singleton_entity("core__aip_executor"),
-            process_executor=self._bootstrap_singleton_entity("core__process_executor"),
-            should_memoize_default=self._bootstrap_singleton_entity(
+        self._core = ExecutionCore(
```

Comment:
It seems to me that we are doing a 2-step initialization? How about having an `initialize` method in `ExecutorCore`? That would make `ExecutorCore` not immutable, but we can add checks that verify the class is initialized before certain properties can be returned.